### ISSUE TITLE
[Bug]: stackplot creates artifacts when height of input is zero

### ISSUE THREAD

**Issue by tlkaufmann** [matplotlib/matplotlib#22393]
### Bug summary

In some cases, when a given input for `stackplot` has zero value entries, thin lines are drawn.

### Code for reproduction

```python
import numpy as np
import matplotlib.pyplot as plt

np.random.seed(42)
x = np.arange(10)
y0 = np.linspace(0, 1, 10)
y1 = np.linspace(0, 1, 10)
y2 = [0, 0, 0.25, 0, 0, 0, 0.25, 0, 0, 0]
y3 = np.linspace(0, 1, 10)


colors=['grey', 'blue', 'red', 'blue']

y = np.stack([y0, y1, y2, y3])
im = plt.stackplot(x, y, colors=colors, lw=0)
```

### Actual outcome

![image](https://user-images.githubusercontent.com/17513791/152529817-df3eb281-a437-43de-93e4-adf2c28d2106.png)

One can see thin grey lines where the red object hast zero height.

### Expected outcome

![image](https://user-images.githubusercontent.com/17513791/152530130-5044f09b-a5b1-4804-8864-441d762368a1.png)

The thin lines are gone

### Additional information

This only happens when the color above and below the empty data has the same color, but there are some use-cases where this is needed (see e.g. [fish plots for evolutionary relationships](https://i.stack.imgur.com/hNl5C.png)).

# Fix
The individual calls to `fill_between` need to have the following parameters:
`where=(stack[i, :] != stack[i+1, :])`, ` interpolate=True` and `edgecolor=color`.

These are the changes made to `matplotlib/lib/matplotlib/stackplot.py`
```python
    # Color between x = 0 and the first array.
    color = ax._get_lines.get_next_color()
    coll = ax.fill_between(x, 0, stack[0, :],
                                      where=(stack[0, :] != 0), interpolate=True, edgecolor=color,
                                     facecolor=color, label=next(labels, None), **kwargs)
    coll.sticky_edges.y[:] = [0]
    r = [coll]

    # Color between array i-1 and array i
    for i in range(len(y) - 1):
        color = ax._get_lines.get_next_color()
        r.append(ax.fill_between(x, stack[i, :], stack[i+1, :],  
                                               where=(stack[i, :] != stack[i+1, :]), interpolate=True, edgecolor=color,
                                               facecolor=color, label=next(labels, None), **kwargs))
```

These changes get rid of the lines but as far as I tested it do not have side effects for other plots. Probably one should check whether edgecolor is set in `**kwargs` as well.

### Matplotlib Version

3.4.3

### Installation

conda

**Comment by jklymak** [comment]
I guess its a question of whether you want the zero-width stack to show up at all or not.  I understand why you want to remove it, but I'm not sure if that is what folks who do stackplots want in general?  (BTW< I'm sure the line is there even if both sides aren't blue, it just may not be as noticeable).

**Comment by tlkaufmann** [comment]
Good point. I also don't know what the standard stackplot user would want. Would be easy to make a parameter like `show_empty` for `stackplot`? If it defaults to False, it won't break any current use-cases.

Either way, I don't think the current behavior is intended because the thin line prevails even if I set `linewidth = 0`. So we are not seeing the outline of the polygon but an artifact of the anti-aliasing. This can be shown as the thin line disappears when setting `aa=False`.

**Comment by jklymak** [comment]
I agree its an anti aliasing artifact.  I guess I'm mildly in favour of your change as fixing a display bug, unless there is some reason to fear it will be a lot slower.  Did you want to open a PR?

**Comment by tlkaufmann** [comment]
I opened a PR for this issue incorporating the fix detailed above: https://github.com/matplotlib/matplotlib/pull/22424

Thanks for your help @jklymak

**Comment by jklymak** [comment]
Let's leave open until the pr is in. Thanks.

**Comment by timhoffm** [comment]
**Carification of root cause:** The underlying problem here is not drawing of a 0-range. You can test by hiding the red area in the above example (`im[2].set_visible(False)`), which does not change the plot.

Instead it is that we have two filled regions (the blue ones in the exmaple) with adjacent edges, and since these polygons are rendered individually there can be antialiasing effects on the edge so that there may be a visual border even though the edges are perfectly aligned.

I'm going to close this as won't fixt for the following reasons:

- This is a fundamental rendering limitation that cannot be handled on our side. - It's the general problem of individual antialiased rendering two polygons that share an edge.
- This will only be visible in stackplots if stack elements i and i+2 have the same color and element i has a 0-region. That should be quite an edge case.
- With #30808 it will be possible to draw finite edges with selected colors. Since edges of adjacent polygons are drawn on top of each other, this will fill any possible gaps in the rendering. So the workaround will be that you define finite edges for the blue areas.
